### DISCUSSION TITLE
Point source on symmetric axis (RZ coord)

### DISCUSSION THREAD

**Discussion by lyyc199586** [idaholab/moose#20087] [Q&A Modules: General] [answered]
Hi all,

I want to model a simple point acoustic source at the symmetric axis. Surely, I can make it in 3D with the point source in the middle but it would be much more efficient to move to an axial symmetric model with a point source at the z-axis. Unfortunately, whenever I try to place a point source at the z-axis the resulting acoustic waves are strange and incorrect. Is there an easy way around this?

Thanks!

**Comment by GiudGiud** [comment]
@somu15 
Could you please post a figure of the problem @lyyc199586  ?

**Reply by lyyc199586** [reply]
sure,
in RZ coord:
![image](https://user-images.githubusercontent.com/20653957/150273465-04c58087-ea16-4015-9a44-414b3766847f.png)
I put a point source on the left side, which is the y-axis, and the acoustic wave becomes very noisy


The same setting in 2D planer gives:
 
![image](https://user-images.githubusercontent.com/20653957/150273625-01cbdc9d-8d87-4b45-b180-2c77d1584852.png)

**Comment by somu15** [comment]
HI @lyyc199586 Would it be possible to share your Moose input files for these two problems? Thanks.

**Comment by lyyc199586** [comment]
This is my input file, to change to a standard coord system, just comment `[Problem]` block.

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  xmin = 0
  xmax = 3
  ymin = 1
  ymax = 4
  nx = 100
  ny = 100
[]

[Problem]
  coord_type = RZ
[]

[Variables]
  [./p]
  [../]
[]

[AuxVariables]
  [./vel_p]
  [../]
  [./accel_p]
  [../]
  [./accel_x]
  [../]
  [./accel_y]
  [../]
[]

[Kernels]
  [./inertia_p]
    type = InertialForce
    variable = 'p'
  [../]
  [./diff_p]
    type = ADMatDiffusion
    variable = 'p'
    diffusivity = 'Diff'
  [../]
[]

[DiracKernels]
  [point_source]
    type = FunctionDiracSource
    variable = p
    function = switch_off
    point = '0.0 1.75 0.0'
  []
[]

[Functions]
  [./switch_off]
    type = ParsedFunction
    value = 'if(t < 0.1, 5e-4, 0)'
  [../]
[]

[Materials]
  [./density]
    type = GenericConstantMaterial
    prop_names = 'density'
    prop_values = '444.44'
  [../]
  [./diff]
    type = ADGenericConstantMaterial
    prop_names = 'Diff'
    prop_values = '1000'
  [../]
[]

[BCs]
  [axial]
    type = DiffusionFluxBC
    variable = p
    boundary = left
  []
[]


[Postprocessors]
  [./p_1]
    type = PointValue
    variable = p
    # point = '0.5 1.01 0.0'
    point = '0 1.75 0'
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm      31                  preonly       lu           1'
  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-8
  # automatic_scaling = true
  end_time = 2.1
  dt = 1.5e-3
  # end_time = 1
  # dt = 1e-3
  [TimeIntegrator]
    type = NewmarkBeta
  []
[]

[Outputs]
  [./csv]
    type = CSV
    delimiter = ','
    file_base = 'pressure_hist'
  [../]
  [./exodus]
    type = Exodus
    interval = 100
    file_base = fluid
  [../]
[]
```

**Reply by jiangwen84** [reply]
@lyyc199586  If you do not put the point source in the center (although for RZ, the point source will become a circular point source), does that solution look reasonable?  It looks like a numerical stability issue, and I am trying to think whether that could be due to the point source at the symmetric plane. 

@somu15 might have some better suggestions for you to try.

**Reply by lyyc199586** [reply]
If the point source is off-axis (in RZ), then it looks fine

**Comment by lynnmunday** [comment]
I think its numerical error.  I changed nx to 1000 and made the timestep 1.5e-4 and the pressure output looks good.... I don't know if its right but it doesn't checkerboard

**Reply by lyyc199586** [reply]
In real simulations a much finer mesh is used, but this problem still exist

**Comment by hugary1995** [comment, marked as answer]
There are two main issues here:

1. The dirac source isn't well defined along the symmetry axis when the coordinate system is rz. In cartesian coordinates the dirac kernel works in the following way: consider the integral `\int_\Omega w s \delta(x-x0) dV`, where `\Omega` is the entire domain, `w` is the test function, `s` is your point source, and `\delta` is the dirac delta function with the spike at `x0`. The integrand gets sifted out and you get `w(x0) s(x0) JxW` in the assembly process. The issue comes up when you do the coordinate transformation for rz, where the integral becomes `\int_\Omega w s \delta(x-x0) 2 \pi r dV`. Now, when this gets sifted out, it becomes zero because `r = 0` along the symmetry axis. To get around this, you either need to use a crazily refined mesh (which can get rid of the noisy waves, but you are still relying on numerical integration error), or change the point source to a well defined volumetric source defined around the neighborhood of that point.
2. ~The `ADMatDiffusion` doesn't account for rz coordinates. The correct residual should be the following one. You can just use the `ADCoefMatDiffusion` we have in raccoon.~

I tested this using the following input file, it works fine without further refinement:
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  xmin = 0
  xmax = 3
  ymin = 1
  ymax = 4
  nx = 100
  ny = 100
[]

[Problem]
  coord_type = RZ
[]

[Variables]
  [p]
  []
[]

[Kernels]
  [inertia_p]
    type = InertialForce
    variable = p
  []
  [diff_p]
    type = ADCoefMatDiffusion
    variable = p
    prop_names = 'D'
  []
  [source_p]
    type = ADCoefMatSource
    variable = p
    prop_names = 's'
    coefficient = -1
  []
[]

[Functions]
  [s_func]
    type = ParsedFunction
    value = 'r:=sqrt(x^2+(y-1.75)^2); if(t<0.1,if(r<0.25,5e-4,0),0)'
  []
[]

[Materials]
  [density]
    type = GenericConstantMaterial
    prop_names = 'density'
    prop_values = '444.44'
  []
  [diff]
    type = ADGenericConstantMaterial
    prop_names = 'D'
    prop_values = '1000'
  []
  [source]
    type = ADGenericFunctionMaterial
    prop_names = 's'
    prop_values = 's_func'
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-8
  end_time = 2.1
  dt = 1.5e-3
  [TimeIntegrator]
    type = NewmarkBeta
  []
[]

[Outputs]
  exodus = true
  print_linear_residuals = false
[]
```

**Reply by lyyc199586** [reply]
Yes, actually `ADCoefMatDiffusion` is being used in real simulations, but this problem still exists.

**Reply by hugary1995** [reply]
The additional term accounting for the lapacian in cylindrical coordinates won't remove the checkerboarding. The volumetric source will.

**Reply by lyyc199586** [reply]
Ok, I will try to use a volumetric source

**Reply by hugary1995** [reply]
Yeah, just make sure the source spans over a few elements. So you might need a little bit of local refinement around that point.

**Reply by lyyc199586** [reply]
Then I just need to change this block to match the original dirac source, right?
```
[Functions]
  [s_func]
    type = ParsedFunction
    value = 'r:=sqrt(x^2+(y-1.75)^2); if(t<0.1,if(r<0.25,5e-4,0),0)'
  []
[]
```

**Reply by hugary1995** [reply]
Right, you might need to do some calculations to make the total power input the same as before.

**Reply by jiangwen84** [reply]
@hugary1995 Initially I thought about the change of lapacian operator in cylindrical coordinate, but we have never add that extra term for any diffusion kernels, including heat conduction. I sorted of convinced my self by looking at the weak form where only gradient term appears and the gradient term remains the same. After you brought this up, I think you are correct and we should start with governing equation and derive the weak form from there. Does that mean all our thermal calculation for RZ is not correct?

**Reply by hugary1995** [reply]
I don't know. I really hope someone proves me wrong on that.

**Reply by jiangwen84** [reply]
@hugary1995 For your first point, I agree that the scalar factor `r` will zero out some contributions from the centerline nodes. However, we still get some nonzero value at two other nodes. Regardless of the accuracy issue, I am curious to know why that would trigger the oscillation in the solution.

**Reply by jiangwen84** [reply]
> I don't know. I really hope someone proves me wrong on that.

@hugary1995  Take a look at https://math.stackexchange.com/questions/2832540/weak-form-of-poisson-equation-in-3d-cylindrical-space-heat-conduction 

I think no extra term is needed for RZ in diffusion kernel.

**Reply by hugary1995** [reply]
> I think no extra term is needed for RZ in diffusion kernel.

Great, I learned something today! I'll fix that in raccoon then.

> For your first point, I agree that the scalar factor r will zero out some contributions from the centerline nodes. However, we still get some nonzero value at two other nodes. Regardless of the accuracy issue, I am curious to know why that would trigger the oscillation in the solution.

Why do you think there are nonzero values? As far as I can tell, the dirac kernels (the test functions, qrule, and everything) are reinit'ed at the points where the dirac delta functions are centered, which agrees with the math. You can check `FEProblemBase::reinitDirac`. I think the oscillations are integration errors from things like `inverse_map` during the reinit.

**Reply by hugary1995** [reply]
Oh no, I just realized that the residual in `DiracKernel::computeResidual` isn't multiplied by the coordinate transformation factor.

**Reply by lyyc199586** [reply]
So `DiracKernel` also need to be modified?

**Comment by lyyc199586** [comment]
So @hugary1995's method is the best alternative so far, but putting a dirac source on the Z-axis is still problematic. I'm just curious is there any way to implement dirac source in RZ

**Comment by Shuvro129** [comment]
Hi, I may have a similar problem at hand. Can you help me?

**Reply by jiangwen84** [reply]
What is your problem?

As @hugary1995 mentioned, the point source is not multiplied by the coordinate transformation factor. This means even with RZ, the point source is treated as a "point", not a "ring".

**Reply by Shuvro129** [reply]
−𝐊 (𝝏^𝟐 𝑼)/(𝝏𝒙^𝟐 )+𝑴 (𝝏^𝟐 𝑼)/(𝝏𝒕^𝟐 )= 𝜹(𝒙,𝒕), I want to solve this problem in infinite 1D domain. That is the left boundary and also the right boundary is infinite. And the Dirac force is implemented on the mid-point. I am confused about how will input the boundary condition.
If you have any ideas tell me.
